### PR TITLE
Add integration tests and prop tests for Osmosis

### LINKED ISSUE #114: Make `cargo deny check licenses` pass
Fails on `downloader` being `LGPL`. It is a dependency in cw-it. Opened https://github.com/apollodao/cw-it/issues/3 to track removing that from cw-it.

### PR DESCRIPTION
**Pull request by apollo-pacman**
- [x] Update CI workflows to match new workspace structure
- [x] Fix so that wasm build only builds the contract 
    - Run with `cargo make wasm`
- [x] Update all the dependencies to not use local versions

Closes #114

### CODE REVIEW

**Review by apollo-bobby — COMMENTED** [review]
I have added an ignore rule for [RUSTSEC-2020-0071](https://rustsec.org/advisories/RUSTSEC-2020-0071), which was incorrectly causing `cargo deny` to report a vulnerability for the `chrono` dependency in `osmosis-std`. Please see my latest commit message for some more details.
We must take care to not ignore this in the future if it resurfaces.

File: cw-dex/tests/osmosis_tests.rs [hunk 1]
```diff
@@ -0,0 +1,378 @@
...
+        let (runner, accs, pool_id, contract_addr) = setup_pool_and_contract(
+            OsmosisPoolType::Basic,
+            vec![1_000_000, 1_000_000],
+            Some(lock_duration),
+        )?;
+        let admin = &accs[0];
+
+        // Query LP token balance
+        let lp_token_denom = format!("gamm/pool/{}", pool_id);
+        let lp_token_balance =
```
Comment (apollo-sturdy) [inline cw-dex/tests/osmosis_tests.rs:194]: Consider renaming this variable as its used later below and it's there a bit unclear which balance it is.

File: .github/workflows/lint-format.yml [hunk 2]
```diff
@@ -2,12 +2,6 @@ name: Linting and formatting
 on:
   pull_request:
   workflow_dispatch:
-    inputs:
-      branch:
-        description: 'Target branch'
-        required: false
-        default: 'develop'
-        type: string
```
Comment (apollo-sturdy) [inline .github/workflows/lint-format.yml:10]: Why remove? Does it work without? How to specify branch when manually triggering workflow without this?

File: .github/workflows/test.yml [hunk 3]
```diff
@@ -39,6 +26,30 @@ jobs:
         uses: actions-rs/cargo@v1
         with:
           command: test
-          args: --locked --all-features
+          args: --locked --lib
```
Comment (apollo-sturdy) [inline .github/workflows/test.yml:29]: Why remove --all-features? Are there no unit tests in feature gated modules we want to test?

File: cw-dex/tests/osmosis_tests.rs [hunk 4]
```diff
@@ -0,0 +1,378 @@
...
+            bank_balance_query(&runner, contract_addr, format!("gamm/pool/{}", pool_id)).unwrap();
+
+        // Assert that LP tokens have been unlocked
+        assert_eq!(lp_token_balance, unlock_amount);
+
+        Ok(())
+    }
+
+    // TODO: For some reason it fails when swap amount is 1. Need to investigate
```
Comment (apollo-sturdy) [inline cw-dex/tests/osmosis_tests.rs:318]: Either fix i this PR or open an issue and remove comment.

File: cw-dex/tests/osmosis_proptests.rs [hunk 5]
```diff
@@ -0,0 +1,177 @@
...
+    setup_pool_and_test_contract(
+        pool_type,
+        initial_liquidity,
+        1_209_600, // Two weeks in seconds
+        1,
+        TEST_CONTRACT_WASM_FILE_PATH,
+    )
+}
+
+fn test_multi_pool_provide_liquidity(
```
Comment (apollo-sturdy) [inline cw-dex/tests/osmosis_proptests.rs:26]: Add doc comment

Comment (apollo-sturdy) [reply cw-dex/tests/osmosis_proptests.rs:26]: Why called multi_pool? What is multi?

File: cw-dex/tests/osmosis_proptests.rs [hunk 6]
```diff
@@ -0,0 +1,177 @@
...
+    let ask_idx = ask_idx % n;
+
+    // If offer and ask are the same, increment ask mod n
+    let ask_idx = if offer_idx == ask_idx {
+        (ask_idx + 1) % n
+    } else {
+        ask_idx
+    };
+
+    // Offer amount cannot be larger than the pool liquidity
```
Comment (apollo-sturdy) [inline cw-dex/tests/osmosis_proptests.rs:112]: Why not? Getting a weird error? :p
Had the same issue in my tests, but on testnet it works fine...

Comment (apollo-sturdy) [reply cw-dex/tests/osmosis_proptests.rs:112]: Maybe clarify the comment for those who aren't aware of this bug(?).

File: cw-dex/tests/osmosis_proptests.rs [hunk 7]
```diff
@@ -0,0 +1,177 @@
...
+        amount: Uint128::from(offer_amount),
+    };
+
+    (pool_liquidity, offer, ask)
+}
+
+proptest! {
+    #![proptest_config(ProptestConfig {
+        // Setting both fork and timeout is redundant since timeout implies
+        // fork, but both are shown for clarity.
```
Comment (apollo-sturdy) [inline cw-dex/tests/osmosis_proptests.rs:134]: Outdated comment? Don't see mention of "timeout" nor "fork" anywhere.

**Review by apollo-sturdy — APPROVED** [review]
Nice work! Approving, but see comments. Could be nice to clarify some stuff.

File: .github/workflows/lint-format.yml (hunk 2 above)
Comment (apollo-bobby) [reply .github/workflows/lint-format.yml:10]: The branch selection is built in nowadays, I didn't realize that when I added it.

File: Cargo.toml [hunk 8]
```diff
@@ -1,41 +1,40 @@
...
-
-[dependencies]
-cosmwasm-schema = "1.1.0"
-cosmwasm-std = {version = "1.1.0", features = ["stargate"]}
+[workspace.dependencies]
+cosmwasm-schema = "1.1.3"
+cosmwasm-std = "1.1.3"
+cosmwasm-storage = "1.1.3"
+cw-storage-plus = "0.15.1"
+cw2 = "0.15.1"
```
[the remark this answers is not in the document]
Comment (apollo-pacman) [reply Cargo.toml:20]: ack

File: cw-dex/Cargo.toml [hunk 9]
```diff
@@ -0,0 +1,51 @@
...
+rustdoc-args = ["--cfg", "docsrs"]
+
+[dependencies]
+cosmwasm-schema = { workspace = true }
+cosmwasm-std = { workspace = true }
+cw-storage-plus = { workspace = true }
+thiserror = { workspace = true }
+cw-asset = { workspace = true }
+cw-utils = "0.16"
+cw20 = "0.16"
```
[the remark this answers is not in the document]
Comment (apollo-pacman) [reply cw-dex/Cargo.toml:27]: ack

File: test-helpers/src/helpers.rs [hunk 10]
```diff
@@ -0,0 +1,49 @@
...
+pub fn provide_liquidity<'a, R: Runner<'a>>(
+    runner: &'a R,
+    contract_addr: String,
+    assets: AssetList,
+    min_out: Uint128,
+    signer: &SigningAccount,
+) {
+    let (funds, _) = separate_natives_and_cw20s(&assets);
+
+    // TODO: Increase allowance for cw20 assets
```
[the remark this answers is not in the document]
Comment (apollo-pacman) [reply test-helpers/src/helpers.rs:65]: https://github.com/apollodao/cw-dex/issues/117

File: test-helpers/Cargo.toml [hunk 11]
```diff
@@ -0,0 +1,34 @@
...
+  --mount type=volume,source=registry_cache,target=/usr/local/cargo/registry \
+  cosmwasm/rust-optimizer:0.12.6
+"""
+
+[dependencies]
+cosmwasm-std = { workspace = true }
+cw-asset = { workspace = true }
+osmosis-testing = { workspace = true }
+osmosis-std = { workspace = true }
+apollo-utils = { git = "https://github.com/apollodao/apollo-utils.git", rev = "bfd1abd8cd9716dccad3e74aeb3704cad9f1f41a" }
```
[the remark this answers is not in the document]
Comment (apollo-pacman) [reply test-helpers/Cargo.toml:32]: it depends on a specific version to have the same version of cw-asset but cw-dex uses a function from a later commit. https://github.com/apollodao/cw-dex/issues/118

File: cw-dex/tests/configs/osmosis.yaml [hunk 12]
```diff
@@ -0,0 +1,31 @@
+folder: "./tests/configs"
```
[the remark this answers is not in the document]
Comment (apollo-pacman) [reply cw-dex/tests/configs/osmosis.yaml:1]: ack

File: test-helpers/src/osmosis.rs [hunk 13]
```diff
@@ -0,0 +1,144 @@
...
+                .unwrap()
+                .data
+                .pool_id
+        }
+        OsmosisPoolType::Balancer { pool_weights } => {
+            gamm.create_balancer_pool(
+                MsgCreateBalancerPool {
+                    sender: signer.address(),
+                    pool_params: Some(PoolParams {
+                        swap_fee: "10000000000000000".to_string(),
```
[the remark this answers is not in the document]
Comment (apollo-pacman) [reply test-helpers/src/osmosis.rs:41]: I dont know and it is not documented in the osmosis proto. I just set the same fee as in the `create_basic_pool` function from `osmosis-testing`

File: test-helpers/src/osmosis.rs [hunk 14]
```diff
@@ -0,0 +1,144 @@
...
+            .unwrap()
+            .data
+            .pool_id
+        }
+        OsmosisPoolType::StableSwap { scaling_factors } => {
+            gamm.create_stable_swap_pool(
+                MsgCreateStableswapPool {
+                    sender: signer.address(),
+                    pool_params: Some(StableSwapPoolParams {
+                        swap_fee: "10000000000000000".to_string(),
```
[the remark this answers is not in the document]
Comment (apollo-pacman) [reply test-helpers/src/osmosis.rs:66]: dito

File: test-helpers/src/osmosis.rs [hunk 15]
```diff
@@ -0,0 +1,144 @@
...
+    lock_id: u64,
+    wasm_file_path: &str,
+) -> RunnerResult<(OsmosisTestApp, Vec<SigningAccount>, u64, String)> {
+    let runner = OsmosisTestApp::new();
+
+    let initial_liquidity = initial_liquidity
+        .iter()
+        .enumerate()
+        .map(|(i, amount)| Coin {
+            denom: format!("denom{}", i),
```
[the remark this answers is not in the document]
Comment (apollo-pacman) [reply test-helpers/src/osmosis.rs:100]: yeah lets refactor when we add test container support

File: .github/workflows/test.yml (hunk 3 above)
Comment (apollo-pacman) [reply .github/workflows/test.yml:29]: ack

File: cw-dex/tests/osmosis_proptests.rs (hunk 5 above)
Comment (apollo-pacman) [reply cw-dex/tests/osmosis_proptests.rs:26]: ack

File: cw-dex/tests/osmosis_proptests.rs (hunk 7 above)
Comment (apollo-pacman) [reply cw-dex/tests/osmosis_proptests.rs:134]: ack

File: cw-dex/tests/osmosis_proptests.rs (hunk 6 above)
Comment (apollo-pacman) [reply cw-dex/tests/osmosis_proptests.rs:112]: yeah getting some error from osmosis

**Comment by apollo-pacman** [comment]
Addressed comments and opened a new PR containing only the osmosis tests here https://github.com/apollodao/cw-dex/pull/116

**Pull request closed without merging**